### PR TITLE
New York Blockchain Week Hackathon GitHub Prize: Write rDAI allocation strategy for Aave

### PR DESCRIPTION
**Pull request by Bobface**
# Aave allocation strategy for the rDai contract
This PR is a submission to the mentioned hackathon and introduces support for the Aave protocol as an allocation strategy.

## New Contracts

- `aave/*`: The official Aave protocol implementation used for testing and interfaces
- `contracts/test/ERC20DetailedMintable`: Mock ERC20 which supports name, symbol, decimals and minting, used for testing
- `contracts/AaveAllocationStrategy`: The implementation of the Aave allocation strategy conforming to the existing interface as used by the Compound allocation strategy

## New Tests
`contracts/test/RDaiWithAave`: Contains tests for rDai with the Aave allocation strategy similar to the existing tests for the Compound allocation strategy

## Implementation details of AaveAllocationStrategy
From the outside, the implementation can be used exactly the same as the existing Compound allocation strategy. However, the implementation has a few differences:

A key difference between Compound and Aave is that Compound models interest accrual by increasing the value of cTokens meaning that if I hold X amount of cTokens I can later swap this amount back to a larger amount of the underlying asset. Aave on the other hand models interest accrual by increasing my balance - the exchange rate between aToken <-> underlying is always 1:1 but if I hold aTokens my balance steadily increases.

For the implementation this means:

`accrueInterest` does not implement any behaviour, it simply returns `true`. There is no explicit interest accrual mechanism in Aave.

`exchangeRateStored` calculates the exchange rate from aToken to underlying as the total balance of aTokens divided by the total invested underlying amount: `rate = (aToken.balanceOf(address(this)) * 10**18) / totalInvested;`. Using this calculation we do not need to modify the existing main rDai contract. 
However, the allocation strategy needs to keep track of the total amount of invested Dai. Thus a new state variable `totalInvested` is introduced which is increased when investing and decreased when redeeming. When investing or redeeming from Aave we need to modify the `totalInvested` variable in a way which will keep the exchange rate unchanged. We can do so by using the formula `totalInvested = (aTotalAfter * totalInvested) / aTotalBefore`. To deduce this formula:
```
// We want to keep the exchange rate while updating the totalInvested. We calculate the newTotalInvested value we need to have the same exchange rate as before

oldExchangeRate = newExchangeRate
oldATokenBalance / oldTotalInvested = newATokenBalance / newTotalInvested      // solve for newTotalInvested
newATokenBalance  / (oldATokenBalance / oldTotalInvested) = newTotalInvested
newTotalInvested = (newATokenBalance * oldTotalInvested) / oldATokenBalance
```

## Kovan deployed addresses

_Note: The Kovan Dai listed below is not the "official" MakerDAO Kovan Dai, it is Aaves own Dai variant with a faucet which is used by Aave on Kovan._)

- Dai: 0xFf795577d9AC8bD7D90Ee22b6C1703490b6512FD
- aDai: 0x58AD4cB396411B691A9AAb6F74545b2C5217FE6a
- Aave Allocation Strategy: 0x446152bbad097508ea6cce34a62d7746c4fc3147
- rDai with Aave Allocation Strategy: 0x487901abd2eb2b5b0c345df795dd357d4638fc26

### Test transactions (see kovan.etherscan.io)
- Mint 100: 0x1800826666972f95986e10725d88f6fbe880547e5df0d9e8ecc1b0498fc71714
- Redeem 50: 0x8ebd814e24a80659492bc1472d29fa43b5ce2a4aea492e94b6e9625b87d45b92

## Example three user interaction
3 user interaction

## Video Demonstration
[rDai with Aave Demo on Youtube](https://www.youtube.com/watch?v=DzfFM-oHaXg)

## Further questions?
Feel free to contact me in this PR or also on Discord (Bobface#9040 - I am on the rDai Discord server)

### CODE REVIEW

**Comment by pi0neerpat** [comment]
Documenting how to test:
0. Build core contracts with yarn build
1. Add contracts_directory: "aave" in truffle-config.js  and run truffle migrate to build the new Aave contracts. 
2. Remove the line you just added in truffle-config.js
3.  Run truffle test test/RTokenWithAave.test.js
(note: yarn test isn't currently injecting mocha dependencies, maybe due to the decentral.ee/web3-test-helpers being used instead of the truffle test suite?

File: packages/contracts/contracts/AaveAllocationStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        // Store the aToken balance of aTokens before deposit
+        uint256 aTotalBefore = aToken.balanceOf(address(this));
+        // Redeem redeemAmount of underlying asset from Aave
+        aToken.redeem(redeemAmount);
+        // Calculate the difference in aToken balance after redeem
+        uint256 aTotalAfter = aToken.balanceOf(address(this));
+        // Check the aToken balance has decreased after redeem
+        require(aTotalAfter <= aTotalBefore, "Aave redeemed negative amount!?");
+        // Update totalInvested
+        totalInvested = (aTotalAfter * totalInvested) / aTotalBefore;
```
Comment (hellwolf) [inline packages/contracts/contracts/AaveAllocationStrategy.sol:90]: This is great, it does the trick.

I really like that in the PR you highlighted the principle:

```
// We want to keep the exchange rate while updating the totalInvested. We calculate the newTotalInvested value we need to have the same exchange rate as before

oldExchangeRate = newExchangeRate
oldATokenBalance / oldTotalInvested = newATokenBalance / newTotalInvested      // solve for newTotalInvested
newATokenBalance  / (oldATokenBalance / oldTotalInvested) = newTotalInvested
newTotalInvested = (newATokenBalance * oldTotalInvested) / oldATokenBalance
```

Would you please take this principle to the comment too, that'd make the code self-explainatory.

File: packages/contracts/test/RTokenWithAave.test.js [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+        }, "invariants: accountStats vs hatStats");
+    }
+
+    it("#0 initial test condition", async () => {
+        assert.equal(wad4human(await rToken.totalSupply.call()), "0.00000");
+        assert.equal(wad4human(await aToken.balanceOf.call(customer1)), "0.00000");
+        assert.equal(wad4human(await token.balanceOf.call(customer1)), "1000.00000");
+    });
+
+    it("#1 normal operations with zero hatter", async () => {
```
Comment (hellwolf) [inline packages/contracts/test/RTokenWithAave.test.js:340]: Good stuff.

Rtoken team will continue from here, and refactor our test suite to test this strategy a bit more

File: packages/contracts/contracts/AaveAllocationStrategy.sol (hunk 1 above)
Comment (Bobface) [reply packages/contracts/contracts/AaveAllocationStrategy.sol:96]: Added the comments to the contract source

**Comment by Bobface** [comment]
@hellwolf I am not sure what makes the pipeline failed. @pi0neerpat suggested it might be caused by using `yarn test` and not `truffle test` because truffle injects mocha dependencies into the tests which are required for `expect`. `yarn test` does not seem to do that for some reason.

**Comment by Bobface** [comment]
@hellwolf @pi0neerpat The tests in the CI are now working. What I have done:

1. Edit `truffle-config.js` to use a glob for the contract directories: `contracts_directory: "./{aave,compound,contracts}/**/*.sol"`. This way all contracts are compiled

2. Import `expect` in `RTokenWithAave.test.js` directly to not have eslint errors: `const expect = require("chai").expect;`

3. Renamed some of Aaves contracts because they had the same name as existing contracts from Compound and rDai which caused the Truffle artifacts to become overwritten. 

The last thing which makes the coverage test fail is that, for some reason, the glob I have mentioned is not expanded by `solidity-coverage` and fails with this error:

`ls: no such file or directory: /home/runner/work/rtoken-monorepo/rtoken-monorepo/packages/contracts/{aave,compound,contracts}/**/*.sol/**/*.sol`

As you can see the glob in raw form is in the path.

**Comment by Bobface** [comment]
I have added a commit which moves the folders into the `contracts` directory as described above and the pipeline is working fine now. If this is not desired I can revert the commit again.

**Pull request merged**